doc/jobs/ci_jobs.rst

```diff
@@ -0,0 +1,188 @@
+  configuration as well as the overlay configuration.
+* **generate_ci_job.py** generates a *CI* nightly job for for each platform
+  and architecture listed in the *CI build file*.
+* **generate_ci_script.py** generates a *shell* script which will run the
+  same tasks as the *CI* job on a local machine.
+
+The build process in detail
+---------------------------
+
+The actual build is performed within a Docker container in order to ensure
+only declared dependencies are available.
+
+The actual build process starts in the script *create_ci_task_generator.py*.
+It generates three Dockerfiles: one to perform the *create-workspace* task to
+populate the workspace and enumerate prerequisites, one to perform the
+*build-and-install* task, and one to perform the *build-and-test* task.
+
+A typical job runs 4 Dockerfiles in all.
+
+Create workspace
+^^^^^^^^^^^^^^^^
+
+This task is performed by the script *create_workspace.py*, the Colcon
+executable and the rosdep executable.
+
+The task performs the following steps:
+
+* Prepare the package sources
+
+  * Fetches the `.repos` file(s)
+  * Fetches the source repositories containing the package(s) to be built
+  * If necessary, switches the branch in those repositories
+
+* Select the packages which should be built, tested, and installed
+
+  * Marks the packages explicitly listed as ignored
+  * If necessary, ignores any packages not selected or in scope
+
+* Enumerate dependencies
+
+  * Lists the dependencies necessary to build and test the remaining ROS
+    packages in the workspace, as well as the dependencies of any ROS packages
+    in the underlay workspace(s) needed to support them.
+  * Uses the rosdep tool to resolve those dependency keys to package names
+
+Build and install
+^^^^^^^^^^^^^^^^^
+
+This task is identical to `the one used by the devel jobs <devel_jobs.rst#Build-and-install>`_.
+
+Build and test
+^^^^^^^^^^^^^^
+
+This task is identical to `the one used by the devel jobs <devel_jobs.rst#Build-and-test>`_.
+
+Known limitations
+^^^^^^^^^^^^^^^^^
+
+System dependency enumeration happens for all ROS packages that are part of the
+non-underlay workspace. This means that any non-ROS packages present in that
+workspace may need their dependencies explicitly called out for inclusion in
+the foundation package list, and also means that a missing dependency may be
+occluded by another package in the workspace correctly declaring the same
+dependency.
+
+Run the *CI* job locally
+------------------------
+
+Example invocation
+^^^^^^^^^^^^^^^^^^
+
+The following commands run the *CI* job for the *ament_cmake_ros* package
+from ROS *Crystal* for Ubuntu *Bionic* *amd64*:
+
+.. code:: sh
+
+  mkdir /tmp/ci_job
+  generate_ci_script.py https://raw.githubusercontent.com/ros2/ros_buildfarm_config/ros2/index.yaml crystal default ubuntu bionic amd64 --packages-select ament_cmake_ros --depth-before 100 > /tmp/ci_job/ci_job_crystal_ament_cmake_ros.sh
```

Comment:
The default is to select only the packages passed to `--packages-select`. The `--depth-before` and `--depth-after` arguments expand the scope away from the specifically selected packages. So a sufficiently large value for `--depth-before` has similar behavior to `--packages-up-to`.

This minimizes the number of available arguments, while still supporting the "build packages from scratch" and the "build a subset of packages in an overlay" scenarios.

For a normal nightly build, this isn't necessary because `--packages-select` isn't specified, so all of the packages get built.

Reply:
Why not use `--packages-up-to ...` instead of `--depth-before 100 --packages-select ...`?

Reply:
That's what I was saying - we can do that, but I'll need to plumb that value through from the script invocation and Jenkins parameters. Not a big deal, but non-zero work. The truth is that I'm just abusing the `scope` verb to obtain that behavior.

Now that I think about it, however, I think that using `--packages-up-to` would cause unexpected behavior when used in combination with `--depth-after`, because that mechanism expands on whatever packages were initially selected, which used to include only the named packages, but now includes all of the packages before them as well.

Reply:
> I think that using `--packages-up-to` would cause unexpected behavior when used in combination with `--depth-after`, because that mechanism expands on whatever packages were initially selected, which used to include only the named packages, but now includes all of the packages before them as well.

Maybe `--depth-after` should be changes into `--packages-above-depth PKG_NAME(s)` to avoid that problem?

Reply:
Do you mean use `--packages-above PKG_NAME(s)` combined with a new argument `--packages-above-depth #`, where the default depth is infinite (current behavior)? I think that would work. If we drop `--depth-before` entirely as proposed in your other comment, we could then drop the `scope` verb.

Reply:
Correlating the `--packages-above` argument with the `--packages-above-depth` argument doesn't sounds appealing to me. E.g. you might want depth 1 above `foo` but also depth 2 above `bar`. So I would lean toward a new separate argument `--packages-above-depth N PKG_NAME(s)` (only usable once for now).

Reply:
Another challenge we'll have here is that we want the OPPOSITE of this. The `scope` verb lets you select the packages you want, and everything else gets marked with `COLCON_IGNORE`. Do you think there is room for a package selection inversion mechanism, or do you see another way?

I feel that depth-after package selection is one of my favorite parts of this feature, and I want to get it right. If we can have that goodness in colcon itself, I'm all for it, I just want to get it right...

Reply:
> Another challenge we'll have here is that we want the OPPOSITE of this.

I am not sure I understand the use case you are referring to. Please elaborate.

> The `scope` verb lets you select the packages you want, and everything else gets marked with `COLCON_IGNORE`. Do you think there is room for a package selection inversion mechanism, or do you see another way?

You can achieve the same by running `colcon list` and remove `colcon list --packages-above-depth` from the result and then create ignore files for the remaining set of packages (if I understand the case correctly).